// inversify/amd/container/container.js
define(["require", "exports", "../bindings/binding", "./lookup", "../planning/planner", "../resolution/resolver", "../syntax/binding_to_syntax", "../utils/serialization", "./container_snapshot", "../utils/guid", "../constants/error_msgs", "../constants/metadata_keys", "../constants/literal_types"], function (require, exports, binding_1, lookup_1, planner_1, resolver_1, binding_to_syntax_1, serialization_1, container_snapshot_1, guid_1, ERROR_MSGS, METADATA_KEY, literal_types_1) {
    "use strict";
    var Container = (function () {
        function Container(containerOptions) {
            if (containerOptions !== undefined) {
                if (typeof containerOptions !== "object") {
                    throw new Error("" + ERROR_MSGS.CONTAINER_OPTIONS_MUST_BE_AN_OBJECT);
                }
                else if (containerOptions.defaultScope === undefined) {
                    throw new Error("" + ERROR_MSGS.CONTAINER_OPTIONS_INVALID_DEFAULT_SCOPE);
                }
                else if (containerOptions.defaultScope !== literal_types_1.BindingScopeEnum.Singleton &&
                    containerOptions.defaultScope !== literal_types_1.BindingScopeEnum.Transient) {
                    throw new Error("" + ERROR_MSGS.CONTAINER_OPTIONS_INVALID_DEFAULT_SCOPE);
                }
                this.options = {
                    defaultScope: containerOptions.defaultScope
                };
            }
            else {
                this.options = {
                    defaultScope: literal_types_1.BindingScopeEnum.Transient
                };
            }
            this.guid = guid_1.guid();
            this._bindingDictionary = new lookup_1.Lookup();
            this._snapshots = [];
            this._middleware = null;
            this.parent = null;
        }
        Container.merge = function (container1, container2) {
            var container = new Container();
            var bindingDictionary = planner_1.getBindingDictionary(container);
            var bindingDictionary1 = planner_1.getBindingDictionary(container1);
            var bindingDictionary2 = planner_1.getBindingDictionary(container2);
            function copyDictionary(origing, destination) {
                origing.traverse(function (key, value) {
                    value.forEach(function (binding) {
                        destination.add(binding.serviceIdentifier, binding.clone());
                    });
                });
            }
            copyDictionary(bindingDictionary1, bindingDictionary);
            copyDictionary(bindingDictionary2, bindingDictionary);
            return container;
        };
        Container.prototype.load = function () {
            var _this = this;
            var modules = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                modules[_i] = arguments[_i];
            }
            var setModuleId = function (bindingToSyntax, moduleId) {
                bindingToSyntax._binding.moduleId = moduleId;
            };
            var getBindFunction = function (moduleId) {
                return function (serviceIdentifier) {
                    var _bind = _this.bind.bind(_this);
                    var bindingToSyntax = _bind(serviceIdentifier);
                    setModuleId(bindingToSyntax, moduleId);
                    return bindingToSyntax;
                };
            };
            var getUnbindFunction = function (moduleId) {
                return function (serviceIdentifier) {
                    var _unbind = _this.unbind.bind(_this);
                    _unbind(serviceIdentifier);
                };
            };
            var getIsboundFunction = function (moduleId) {
                return function (serviceIdentifier) {
                    var _isBound = _this.isBound.bind(_this);
                    return _isBound(serviceIdentifier);
                };
            };
            var getRebindFunction = function (moduleId) {
                return function (serviceIdentifier) {
                    var _rebind = _this.rebind.bind(_this);
                    var bindingToSyntax = _rebind(serviceIdentifier);
                    setModuleId(bindingToSyntax, moduleId);
                    return bindingToSyntax;
                };
            };
            modules.forEach(function (module) {
                var bindFunction = getBindFunction(module.guid);
                var unbindFunction = getUnbindFunction(module.guid);
                var isboundFunction = getIsboundFunction(module.guid);
                var rebindFunction = getRebindFunction(module.guid);
                module.registry(bindFunction, unbindFunction, isboundFunction, rebindFunction);
            });
        };
        Container.prototype.unload = function () {
            var _this = this;
            var modules = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                modules[_i] = arguments[_i];
            }
            var conditionFactory = function (expected) { return function (item) {
                return item.moduleId === expected;
            }; };
            modules.forEach(function (module) {
                var condition = conditionFactory(module.guid);
                _this._bindingDictionary.removeByCondition(condition);
            });
        };
        Container.prototype.bind = function (serviceIdentifier) {
            var defaultScope = literal_types_1.BindingScopeEnum.Transient;
            defaultScope = (this.options.defaultScope === defaultScope) ? defaultScope : literal_types_1.BindingScopeEnum.Singleton;
            var binding = new binding_1.Binding(serviceIdentifier, defaultScope);
            this._bindingDictionary.add(serviceIdentifier, binding);
            return new binding_to_syntax_1.BindingToSyntax(binding);
        };
        Container.prototype.rebind = function (serviceIdentifier) {
            this.unbind(serviceIdentifier);
            return this.bind(serviceIdentifier);
        };
        Container.prototype.unbind = function (serviceIdentifier) {
            try {
                this._bindingDictionary.remove(serviceIdentifier);
            }
            catch (e) {
                throw new Error(ERROR_MSGS.CANNOT_UNBIND + " " + serialization_1.getServiceIdentifierAsString(serviceIdentifier));
            }
        };
        Container.prototype.unbindAll = function () {
            this._bindingDictionary = new lookup_1.Lookup();
        };
        Container.prototype.isBound = function (serviceIdentifier) {
            return this._bindingDictionary.hasKey(serviceIdentifier);
        };
        Container.prototype.isBoundNamed = function (serviceIdentifier, named) {
            return this.isBoundTagged(serviceIdentifier, METADATA_KEY.NAMED_TAG, named);
        };
        Container.prototype.isBoundTagged = function (serviceIdentifier, key, value) {
            var bindings = this._bindingDictionary.get(serviceIdentifier);
            var request = planner_1.createMockRequest(this, serviceIdentifier, key, value);
            return bindings.some(function (b) { return b.constraint(request); });
        };
        Container.prototype.snapshot = function () {
            this._snapshots.push(container_snapshot_1.ContainerSnapshot.of(this._bindingDictionary.clone(), this._middleware));
        };
        Container.prototype.restore = function () {
            var snapshot = this._snapshots.pop();
            if (snapshot === undefined) {
                throw new Error(ERROR_MSGS.NO_MORE_SNAPSHOTS_AVAILABLE);
            }
            this._bindingDictionary = snapshot.bindings;
            this._middleware = snapshot.middleware;
        };
        Container.prototype.createChild = function () {
            var child = new Container();
            child.parent = this;
            return child;
        };
        Container.prototype.applyMiddleware = function () {
            var middlewares = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                middlewares[_i] = arguments[_i];
            }
            var initial = (this._middleware) ? this._middleware : this._planAndResolve();
            this._middleware = middlewares.reduce(function (prev, curr) {
                return curr(prev);
            }, initial);
        };
        Container.prototype.get = function (serviceIdentifier) {
            return this._get(false, false, literal_types_1.TargetTypeEnum.Variable, serviceIdentifier);
        };
        Container.prototype.getTagged = function (serviceIdentifier, key, value) {
            return this._get(false, false, literal_types_1.TargetTypeEnum.Variable, serviceIdentifier, key, value);
        };
        Container.prototype.getNamed = function (serviceIdentifier, named) {
            return this.getTagged(serviceIdentifier, METADATA_KEY.NAMED_TAG, named);
        };
        Container.prototype.getAll = function (serviceIdentifier) {
            return this._get(true, true, literal_types_1.TargetTypeEnum.Variable, serviceIdentifier);
        };
        Container.prototype.getAllTagged = function (serviceIdentifier, key, value) {
            return this._get(false, true, literal_types_1.TargetTypeEnum.Variable, serviceIdentifier, key, value);
        };
        Container.prototype.getAllNamed = function (serviceIdentifier, named) {
            return this.getAllTagged(serviceIdentifier, METADATA_KEY.NAMED_TAG, named);
        };
        Container.prototype._get = function (avoidConstraints, isMultiInject, targetType, serviceIdentifier, key, value) {
            var result = null;
            var defaultArgs = {
                avoidConstraints: avoidConstraints,
                contextInterceptor: function (context) { return context; },
                isMultiInject: isMultiInject,
                key: key,
                serviceIdentifier: serviceIdentifier,
                targetType: targetType,
                value: value
            };
            if (this._middleware) {
                result = this._middleware(defaultArgs);
                if (result === undefined || result === null) {
                    throw new Error(ERROR_MSGS.INVALID_MIDDLEWARE_RETURN);
                }
            }
            else {
                result = this._planAndResolve()(defaultArgs);
            }
            return result;
        };
        Container.prototype._planAndResolve = function () {
            var _this = this;
            return function (args) {
                var context = planner_1.plan(_this, args.isMultiInject, args.targetType, args.serviceIdentifier, args.key, args.value, args.avoidConstraints);
                context = args.contextInterceptor(context);
                var result = resolver_1.resolve(context);
                return result;
            };
        };
        return Container;
    }());
    exports.Container = Container;
});
